### DISCUSSION TITLE
TypeError: Cannot read property 'initialized' of undefined caused by a custom node

### DISCUSSION THREAD

**Discussion by SciManJD** [bcakmakoglu/vue-flow#874] [Q&A] [answered]
I created a custom node. Everything works fine, but in the console of the browser I see 3 errors like this:

runtime-core.esm-bundler.js:226 TypeError: Cannot read property 'initialized' of undefined
    at vue-flow-core.mjs:1103
    at P (vue-flow-core.mjs:7)
    at z (vue-flow-core.mjs:116)
    at W (vue-flow-core.mjs:139)
    at setup (vue-flow-core.mjs:1103)
    at ft (runtime-core.esm-bundler.js:155)
    at ki (runtime-core.esm-bundler.js:7160)
    at Oi (runtime-core.esm-bundler.js:7115)
    at P (runtime-core.esm-bundler.js:5468)
    at L (runtime-core.esm-bundler.js:5443)

I'm wandering what could be the reason?

**Comment by bcakmakoglu** [comment]
You'll need to give me more information than that, this is telling me nothing 😄 
What's your custom node look like, when and where is this issue appearing?
Provide as much info as you can, please.

**Reply by SciManJD** [reply]
This is the complete code of my new node:

```
<template>
    <div>
        <div id="flow-chart-sum">+</div>
        <Handle id="a" type="target" :position="Position.Top" :style="{filter: 'invert(100%)', left: '16px', top: '0px'}"/>
        <Handle id="b" type="target" :position="Position.Left" :style="{filter: 'invert(100%)', left: '0px', top: '16px'}"/>
        <Handle id="c" type="source" :position="Position.Right" :style="{filter: 'invert(100%)', right: '0px', top: '16px'}"/>
    </div>
</template>
  
<script>
    import {Position, Handle} from '@vue-flow/core';

    export default {
        name: 'VueFlowSum',
        components: {
            Handle
        },
        data() {
            return {
                Position: Position
            }
        }
    }
</script>

<style>
    #flow-chart-sum {
        background: var(--g-primary-color);
        border: 5px solid var(--g-primary-color-darker);
        border-radius: 50%;
        width: 32px;
        padding: 0 8px;
    }
</style>

```

Then I use it like this:

```
<template>
    <VueFlow v-model="elements" fit-view-on-init :default-edge-options="{type: 'smoothstep'}" :node-types="nodeTypes">
        <template>
            <VueFlowSum/>
        </template>
    </VueFlow>
</template>

<script>
    import {VueFlow, Position, useVueFlow} from '@vue-flow/core'
    import VueFlowSum from '@/components/VueFlowSum.vue'
    import {markRaw} from 'vue';
    export default {
        name: 'Component',
        components: {
            VueFlow,
            VueFlowSum
        },
        data() {
            return {
                elements: [...],
                nodeTypes:  {
                    sum: markRaw(VueFlowSum),
                }
            }
        }
    }
</script>

```

**Reply by bcakmakoglu** [reply]
What is this supposed to do?

```js
        data() {
            return {
                Position: Position
            }
        }
```

`Position` is an enum so you should be using it like `position: Position.Left` and not `Position: Position`

Next point:
What is this supposed to do?

```html
 <template>
            <VueFlowSum/>
</template>
```

You're just passing a node down to the default slot but the default slot is not a slot for nodes.

If you want to use a custom node you'll have to use it like this

```html
<!-- `custom` can be whatever *type* you chose as your custom node type, in this example I'll just use `type: 'custom'`
<template #node-custom="nodeProps">
	<CustomNode v-bind="nodeProps" />
</template>
```

Or if you want to use the `nodeTypes` object that's also ok but don't do both.

**Reply by SciManJD** [reply]
About the first note: 
```
        data() {
            return {
                Position: Position
            }
        }
```
I cannot compile the code without this. I get error message: 
` error  'Position' is defined but never used  no-unused-vars`
If make this assignment, Position becomes a part of the local data and I can access it with this.Position or directly in the template. Probably this is a lack of knowledge from me, but this is how I could solve the problem. That way all works fine.

About your second note... yes, you are right. I made many experiments trying to learn how to use the component and some code was left, but it should be removed. I will fix this. Sorry for the spam...

**Reply by bcakmakoglu** [reply]
No worries 😄 
Just confused about the code so I'm asking as that could lead to potential problems ^^

When you get the error you mentioned first, can you show me which part of the code it relates to? There should be a console error where you can open the debugger and see at what line in the library code this happened. Maybe that can help me find the cause 😄

**Reply by SciManJD** [reply]
I cannot debug, because the error comes during the build. I cannot run the application. It seems that this enum cannot be found when it is global defined. Probably this is again a lack of knowledge (sorry, I quite new in the programming with Vue) and haven't programmed correctly something.
And another question... I fixed the code concerning the including of my custom node. Now it looks so (I don't use ant props just for now):

```
             <VueFlow v-model="elements" fit-view-on-init :default-edge-options="{type: 'smoothstep'}">
                <template #node-sum>
                    <VueFlowSum/>
                </template>
            </VueFlow>

```
I can compile and run the application, but now my custom node is missing.

**Reply by bcakmakoglu** [reply]
Did you define your node type as `sum`?

To use this slot you need to define the type in a node, for example

```js
{
	id: '1',
	position: { x: 0, y: 0 },
	type: 'sum' // <--- you have to assign the type here
}
```

**Reply by SciManJD** [reply]
Sorry, forget for the question for including of the custom node. It works fine. But I have commented out the line "Position: Position" in the node component to check what will happen and the component couldn't be compiled.

**Reply by SciManJD** [reply]
Yes, my node is called 'sum' and I defined the type like this when I want to use it:
```
...
{id: 'n-5', type: 'sum', targetPosition: Position.Top, position: {x: 450, y: 284}},
...
```
But as I said I found the problem.

**Reply by SciManJD** [reply, marked as answer]
I forgot to tell, but also the errors now disappeared. So the problem is actually solved.
